cmd/podman/images/save.go

```diff
@@ -95,7 +99,16 @@ func save(cmd *cobra.Command, args []string) error {
 		if terminal.IsTerminal(int(fi.Fd())) {
 			return errors.Errorf("refusing to save to terminal. Use -o flag or redirect")
 		}
-		saveOpts.Output = "/dev/stdout"
+		pipePath, cleanup, err = setupPipe()
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err = cleanup(); err != nil {
```

Comment:
AFAICS `cleanup()` blocks until the pipe is opened and closed, which may never happen on some of the error paths, and this would just hang.

Maybe use `context.Context` to terminate that wait, or just expose `errc` and have this function inject a fallback “wait canceled” error that is then ignored in favor of the real cause.

Reply:
@mtrmac PTAL

Reply:
how to use context.Context to terminate?

Reply:
The way the `cleanup` function now returns a channel, a separate termination context might not be necessary:
```go
func save(…) (finalErr error) {
succeeded := false
…
if /* use pipe */ {
    pipePath, cleanup, err := setupPipe()
    …
    if cleanup != nil {
        defer func() {
            channel := cleanup()
            if succeeded {
                writeErr := <- channel
                if writeErr != nil && finalErr == nil { finalErr = writeErr }
            }
        }
   }
}
…
err := registry.ImageEngine().Save(context.Background(), args[0], tags, saveOpts)
if err == nil { succeeded = true }
return err
}
```